test/functional/test_framework/socks5.py

```diff
@@ -117,6 +135,38 @@ def handle(self):
             cmdin = Socks5Command(cmd, atyp, addr, port, username, password)
             self.serv.queue.put(cmdin)
             logger.debug('Proxy: %s', cmdin)
+
+            requested_to_addr = addr.decode("utf-8")
+            requested_to = format_addr_port(requested_to_addr, port)
+
+            if self.serv.conf.destinations_factory is not None:
+                dest = self.serv.conf.destinations_factory(requested_to_addr, port)
+                if dest is not None:
+                    logger.debug(f"Serving connection to {requested_to}, will redirect it to "
+                                 f"{dest['actual_to_addr']}:{dest['actual_to_port']} instead")
+                    with socket.create_connection((dest["actual_to_addr"], dest["actual_to_port"])) as conn_to:
+                        self.conn.setblocking(False)
+                        conn_to.setblocking(False)
+                        sockets = [self.conn, conn_to]
+                        done = False
+                        while not done:
+                            rlist, _, xlist = select.select(sockets, [], sockets)
+                            if len(xlist) > 0:
+                                raise IOError("Exceptional condition on socket")
+                            for s in rlist:
+                                data = s.recv(4096)
+                                if data is None or len(data) == 0:
+                                    done = True
+                                    break
+                                if s == self.conn:
+                                    conn_to.sendall(data)
```

Comment:
i don't think `sendall` can be guaranteed with non-blocking sockets, it will fail if it gets EAGAIN
https://stackoverflow.com/questions/6240737/python-socket-sendall-function

Reply:
Ok, [`sendall()`](https://docs.python.org/3/library/socket.html#socket.socket.sendall) doc says "On error, an exception is raised, and there is no way to determine how much data, if any, was successfully sent" so this is useless for us. The doc says as well that since Python 3.5 the method will retry if the syscall is interrupted, but mentions nothing about `EAGAIN` or `EWOULDBLOCK`.

So I followed https://docs.python.org/3/howto/sockets.html#using-a-socket and used the basic `send()` with a manual retry loop, like in that example.

Diving further into Python source code, [`sendall()`](https://github.com/python/cpython/blob/b454662921fd3a1fc27169e91aca03aadea08817/Modules/socketmodule.c#L4418) [calls](https://github.com/python/cpython/blob/b454662921fd3a1fc27169e91aca03aadea08817/Modules/socketmodule.c#L4461) [ `sock_call_ex()`](https://github.com/python/cpython/blob/b454662921fd3a1fc27169e91aca03aadea08817/Modules/socketmodule.c#L886) which is [retrying](https://github.com/python/cpython/blob/b454662921fd3a1fc27169e91aca03aadea08817/Modules/socketmodule.c#L994-L1004) on `EWOULDBLOCK` and `EAGAIN`. This is not documented explicitly anywhere, so I decided to ignore it and go with the manual retry loop. Also, `send()` (in addition to `sendall()`) ends up with that retry, so maybe the manual retry loop will always loop just once.